app/src/main/java/org/oppia/android/app/onboarding/OnboardingFragmentPresenter.kt

```diff
@@ -36,13 +84,197 @@ class OnboardingFragmentPresenter @Inject constructor(
         appLanguageResourceHandler.getStringInLocale(R.string.app_name)
       )
 
-      onboardingLanguageLetsGoButton.setOnClickListener {
-        val intent =
-          OnboardingProfileTypeActivity.createOnboardingProfileTypeActivityIntent(activity)
-        fragment.startActivity(intent)
+      onboardingAppLanguageViewModel.supportedAppLanguagesList.observe(
+        fragment,
+        { languagesList ->
+          val adapter = ArrayAdapter(
+            fragment.requireContext(),
+            R.layout.onboarding_language_dropdown_item,
+            R.id.onboarding_language_text_view,
+            languagesList
+          )
+          onboardingLanguageDropdown.setAdapter(adapter)
+        }
+      )
+
+      onboardingAppLanguageViewModel.languageSelectionLiveData.observe(
+        fragment,
+        { language ->
+          selectedLanguage = language
+          onboardingLanguageDropdown.setText(selectedLanguage, false)
+        }
+      )
+
+      onboardingLanguageDropdown.apply {
+        setRawInputType(EditorInfo.TYPE_NULL)
+
+        onItemClickListener =
+          AdapterView.OnItemClickListener { _, _, position, _ ->
+            adapter.getItem(position).let { selectedItem ->
+              if (selectedItem != null) {
+                selectedLanguage = selectedItem as String
+                onboardingAppLanguageViewModel.setSelectedLanguageDisplayName(selectedLanguage)
+              }
+            }
+          }
       }
+
+      onboardingLanguageLetsGoButton.setOnClickListener { updateSelectedLanguage(selectedLanguage) }
     }
 
     return binding.root
   }
+
+  private fun updateSelectedLanguage(selectedLanguage: String) {
+    val oppiaLanguage = appLanguageResourceHandler.getOppiaLanguageFromDisplayName(selectedLanguage)
+    val selection = AppLanguageSelection.newBuilder().setSelectedLanguage(oppiaLanguage).build()
+    translationController.updateAppLanguage(profileId, selection).toLiveData()
+      .observe(
+        fragment,
+        { result ->
+          when (result) {
+            is AsyncResult.Success -> {
+              val intent =
+                OnboardingProfileTypeActivity.createOnboardingProfileTypeActivityIntent(activity)
+              intent.decorateWithUserProfileId(profileId)
+              fragment.startActivity(intent)
+            }
+            is AsyncResult.Failure -> oppiaLogger.e(
+              "OnboardingFragment",
+              "Failed to set AppLanguageSelection",
+              result.error
+            )
```

Comment:
Should we have some sort of fail-safe here? Otherwise the user will potentially get stuck in this flow and be unable to proceed.

Reply:
I opted to finish the activity/close the app.

Reply:
Hmm there's not a perfect solution, but perhaps we should crash rather than just close the app (since then it at least gets special treatment from analytics systems and hopefully gets escalated to us more rapidly than filtering among non-fatal error logs).

I think you can just wrap ``result.error`` in a new exception with context and then throw it here. Definitely a good scenario to add a test for if that's possible.

Reply:
Modified flow to allow progres regardless of saving outcome.